src/platform/Linux/DeviceInfoProviderImpl.cpp

```diff
@@ -0,0 +1,129 @@
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <lib/core/CHIPTLV.h>
+#include <lib/support/CHIPMemString.h>
+#include <lib/support/CodeUtils.h>
+#include <platform/Linux/DeviceInfoProviderImpl.h>
+#include <platform/internal/CHIPDeviceLayerInternal.h>
+
+#include <stdlib.h>
+#include <string.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+DeviceInfoProviderImpl & DeviceInfoProviderImpl::GetDefaultInstance()
+{
+    static DeviceInfoProviderImpl sInstance;
+    return sInstance;
+}
+
+CHIP_ERROR DeviceInfoProviderImpl::SetUserLabelCount(EndpointId endpoint, size_t val)
+{
+    // TODO:: store the user label count.
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR DeviceInfoProviderImpl::GetUserLabelCount(EndpointId endpoint, size_t & val)
+{
+    // TODO:: read the user label count. temporarily return the size of hardcoded labelList.
+    val = 4;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceInfoProviderImpl::SetUserLabelAt(EndpointId endpoint, size_t index, const UserLabelType & userLabel)
+{
+    // TODO:: store the user labelList, and read back stored user labelList if it has been set.
+    // Add yaml test to verify this.
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR DeviceInfoProviderImpl::GetUserLabelAt(EndpointId endpoint, size_t index, UserLabelType & userLabel)
+{
+    // TODO:: get the user labelList from persistent storage, temporarily, use the following
+    // hardcoded labelList on all endpoints.
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    const char * labelPtr = nullptr;
+    const char * valuePtr = nullptr;
+
+    switch (index)
+    {
+    case 0:
+        labelPtr = "room";
+        valuePtr = "bedroom 2";
+        break;
+    case 1:
+        labelPtr = "orientation";
+        valuePtr = "North";
+        break;
+    case 2:
+        labelPtr = "floor";
+        valuePtr = "2";
+        break;
+    case 3:
+        labelPtr = "direction";
+        valuePtr = "up";
+        break;
+    default:
+        err = CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND;
+        break;
+    }
+
+    if (err == CHIP_NO_ERROR)
+    {
+        VerifyOrReturnError(std::strlen(labelPtr) <= kMaxLabelLength, CHIP_ERROR_BUFFER_TOO_SMALL);
+        VerifyOrReturnError(std::strlen(valuePtr) <= kMaxLabelValueLength, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+        Platform::CopyString(mUserLabelBuf, kMaxLabelLength, labelPtr);
+        Platform::CopyString(mUserValueBuf, kMaxLabelValueLength, valuePtr);
```

Comment:
The buffer should be in the iterator, so if there are multiple iterators being used they don't stomp on each other.